### ISSUE TITLE
Appsody deploy always triggers build, even when there are no code changes

### ISSUE THREAD

**Issue by nastacio** [appsody/appsody#980]
**Describe the bug**
If you call `appsody build` and then `appsody deploy` in sequence, without any code modifications in between, `appsody deploy` will rebuild the whole application again before deployment.

**To Reproduce**
1. Create a new application with 
```appsody init incubator/java-openliberty```
 (same result with `java-spring-boot2` and `nodejs-express`)

2. Deploy the application to a target container with 
```
kubectl create namespace cloudlab
appsody deploy -n cloudlab
```

3. Start the deployment again:
```
appsody deploy -n cloudlab
```

**Expected behavior**
If no code was modified since the last build, then the code should not be rebuilt.

**Actual behavior**
`appsody deploy...` triggers a full docker build of the code before initiating the deployment.

**Environment Details (please complete the following information):**
 - OS: [e.g. iOS] All

**If applicable please specify:**
 - CLI version: appsody 0.5.9
 - Stack you are using: java-openliberty, java,spring-boot2, nodejs-express

**Screenshots**

**Additional context**
I think the problem is the absence of `app-deploy.yaml` in the `.dockerignore` file for each stack.

One could argue this is a problem with the stacks, but it is probably best to fix this globally in the appsody cli (maybe appending the `app-deploy.yaml` to `.dockerignore` before builds) .

**Comment by neeraj-laad** [comment]
If you want to do deploy without a build, please use this flag while running `appsody deploy`:

--no-build                 Deploys the application without building a new image or modifying the deployment manifest file.

Just to explain the behaviour:

When `appsody build` is run
- start the appsody container
- extract out stack code & dependencies + Dockerile + user's app & dependencies out
- build the application image using the extracted assets
- clean up the extracted folder

Even though the application files haven't changed, they are extracted again and for the build engine, these are new files every time - resulting in not using the previous layer. 

To automatically do this would need us to not clean up extracted folder and to somehow track if changes have been made since last extract etc. which could be messy.

**Comment by neeraj-laad** [comment]
Closing this as working as designed. 

@nastacio please feel free to re-open if you believe Appsody could do something better here.

**Comment by nastacio** [comment]
@neeraj-laad , apparently I cannot reopen the issue (no button for it) .

I never assumed this was a matter of design, because `appsody build` has no business creating or modifying `app-deploy.yaml`.

How can `appsody build` populate something like `applicationImage` when the caller has no idea where the image will be deployed, or in other words, what the deployer will use as the `--pull-url` parameter.

If you take OpenShift as an example, the `push-url` will almost always be different than `pull-url`, so you must specify `pull-url` in `app-deploy.yaml` and adding --no-build will prevent an update to `app-deploy.yaml`

In short, the current design virtually forces the deployer to *not* add `--no-build` flag unless he is lucky enough to have the correct applicationImage value in `app-deploy.yaml`.
